### PR TITLE
BUG: Fixes einsum broadcasting bug when optimize=True

### LINKED ISSUE #10930: PR #10559  failed to fix einsum (optimize=True) broadcasting bug
#10559 introduced the [following code](https://github.com/numpy/numpy/blob/a9e9343ecf5a6564eb09f1d7078c407db2ff7d9e/numpy/core/einsumfunc.py#L1107) to prevent dispatching `numpy.tensordot` in a case where einsum was broadcasting over a singleton dimension.
```python
        # Handle broadcasting vs BLAS cases
        if blas:
            # Checks have already been handled
            input_str, results_index = einsum_str.split('->')
            input_left, input_right = input_str.split(',')
            if 1 in tmp_operands[0] or 1 in tmp_operands[1]:
                left_dims = {dim: size for dim, size in
                             zip(input_left, tmp_operands[0].shape)}
                right_dims = {dim: size for dim, size in
                              zip(input_right, tmp_operands[1].shape)}
                # If dims do not match we are broadcasting, BLAS off
                if any(left_dims[ind] != right_dims[ind] for ind in idx_rm):
                    blas = False
```
However, this checks to see if `1` occurs within the operand array itself rather than the shape of the operand. Incidentally, this likely produced a nasty performance regression.

Thus the line

`if 1 in tmp_operands[0] or 1 in tmp_operands[1]` 

should be 

`if 1 in tmp_operands[0].shape or 1 in tmp_operands[1].shape`

This wasn't caught by the [unit test](https://github.com/numpy/numpy/blob/a9e9343ecf5a6564eb09f1d7078c407db2ff7d9e/numpy/core/tests/test_einsum.py#L486) because arrays of ones were used 🌌

This leads to the following behavior:

```python
>>> x = np.array([0., 1., 0.])  # contains 1, no blas
>>> y = np.array([0.0])
>>> np.einsum("i,i", x, y, optimize=True)
0.
```

```python
>>> x = np.array([0., -1., 0.])  # doesn't contain 1, yes blas
>>> y = np.array([0.0])
>>> np.einsum("i,i", x, y, optimize=True)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-184-b0dcea8eedea> in <module>()
      1 x = np.array([0., -1., 0.])
      2 y = np.array([0.0])
----> 3 np.einsum("i,i", x, y, optimize=True)

c:\anaconda\envs\py36\lib\site-packages\numpy\core\einsumfunc.py in einsum(*operands, **kwargs)
   1132 
   1133             # Contract!
-> 1134             new_view = tensordot(*tmp_operands, axes=(tuple(left_pos), tuple(right_pos)))
   1135 
   1136             # Build a new view if needed

c:\anaconda\envs\py36\lib\site-packages\numpy\core\numeric.py in tensordot(a, b, axes)
   1281                 axes_b[k] += ndb
   1282     if not equal:
-> 1283         raise ValueError("shape-mismatch for sum")
   1284 
   1285     # Move the axes to sum over to the end of "a"

ValueError: shape-mismatch for sum
```

### PR DESCRIPTION
**Pull request by rsokl**
Closes #10930.

### CODE REVIEW

File: numpy/core/tests/test_einsum.py [hunk 1]
```diff
@@ -490,17 +490,25 @@ def check_einsum_sums(self, dtype, do_opt=False):
                            np.einsum('ij,ij->j', p, q, optimize=False))
         assert_array_equal(np.einsum('ij,ij->j', p, q, optimize=True),
                            [10.] * 2)
-
-        p = np.ones((1, 5))
-        q = np.ones((5, 5))
+        
+        # a blas-compatible contraction broadcasting case which was failing
```
Comment (charris) [inline numpy/core/tests/test_einsum.py]: There is trailing whitespace on these two lines.

Comment (charris) [reply numpy/core/tests/test_einsum.py]: I'll fix it.

File: numpy/core/tests/test_einsum.py [hunk 2]
```diff
@@ -491,16 +491,24 @@ def check_einsum_sums(self, dtype, do_opt=False):
...
-        q = np.ones((5, 5))
+        # a blas-compatible contraction broadcasting case which was failing
+        # for optimize=True (ticket #10930)
+        x = np.array([0., 0.])
+        y = np.array([0.])
+        assert_array_equal(np.einsum("i, i", x, y, optimize=False), 0.)
+        assert_array_equal(np.einsum("i, i", x, y, optimize=True), 0.)
+
+        # all-ones array was bypassing bug (ticket #10930)
+        p = np.ones((1, 5)) / 2
```
Comment (charris) [inline numpy/core/tests/test_einsum.py]: Curious why you wanted floats here.

Comment (charris) [reply numpy/core/tests/test_einsum.py]: NVM, I see you added a note.

**Review by charris — APPROVED** [review]

**Comment by charris** [comment]
LGTM, I'll wait on @dgasmith .

File: numpy/core/tests/test_einsum.py [hunk 3]
```diff
@@ -491,16 +491,24 @@ def check_einsum_sums(self, dtype, do_opt=False):
         assert_array_equal(np.einsum('ij,ij->j', p, q, optimize=True),
                            [10.] * 2)
 
-        p = np.ones((1, 5))
-        q = np.ones((5, 5))
+        # a blas-compatible contraction broadcasting case which was failing
+        # for optimize=True (ticket #10930)
+        x = np.array([0., 0.])
```
Comment (dgasmith) [inline numpy/core/tests/test_einsum.py]: Can we use array of say 5s just incase we hit some sort of weird bug.

**Review by dgasmith — APPROVED** [review]
LGTM, thanks for the PR.

File: numpy/core/tests/test_einsum.py (hunk 3 above)
Comment (rsokl) [reply numpy/core/tests/test_einsum.py]: Sure, I'll have to wait until I'm home from work.

**Comment by rsokl** [comment]
@dgasmith this should be good to go now (updated the test as you requested)

**Comment by charris** [comment]
Thanks @rsokl . For next time, might take a look at the howto for commit messages at `doc/source/dev/gitwash/development_workflow.rst`.

**Comment by rsokl** [comment]
Ah thanks for pointing me to that. I wasn't aware of it. Sorry about that.

**Pull request merged**